app/src/main/java/com/lagradost/cloudstream3/plugins/PluginManager.kt

```diff
@@ -265,16 +267,102 @@ object PluginManager {
         }
 
         main {
-            createNotification(activity, updatedPlugins)
+            createNotification(activity, updatedPlugins, R.string.plugins_updated)
         }
 
-       // ioSafe {
+        // ioSafe {
             afterPluginsLoadedEvent.invoke(true)
-       // }
+        // }
 
         Log.i(TAG, "Plugin update done!")
     }
 
+    /**
+     * Automatically download plugins not yet existing on local
+     * 1. Gets all online data from online plugins repo
+     * 2. Fetch all not downloaded plugins
+     * 3. Download them and reload plugins
+     **/
+    fun downloadNotExistingPluginsAndLoad(activity: Activity, isLoaded: Boolean) {
+        // Load all plugins as fast as possible, if its not loaded yet.
+        if (!isLoaded) {
+            loadAllOnlinePlugins(activity)
```

Comment:
Not a big fan of making the method doing two things at once like this :pensive: 
Why not move the code in this if statement to the MainActivity code? That would be more legible imo

Reply:
I am a guilty of doing it myself which is probably why you did, but I've had real issues debugging the race conditions afterwards :skull:

Reply:
The method runs when auto plugin is enabled (cause it is run inside it). This is just a checker to not run it when its already executed on the auto plugin update method. If not, it runs it like how it is before

Reply:
I could move the code, and have it run before checking for plugin updates

Reply:
Why not something like this?
```kt
                    if (settingsManager.getBoolean(
                            getString(R.string.auto_update_plugins_key),
                            true
                        )
                    ) {
                        PluginManager.updateAllOnlinePluginsAndLoadThem(this@MainActivity)
                    } else {
                        PluginManager.loadAllOnlinePlugins(this@MainActivity)
                    }
                    // Now all online plugins are loaded
                    if (....) PluginManager.downloadNotExistingPluginsAndLoad(this@MainActivity)
```
All plugins are loaded before the downloadNotExistingPluginsAndLoad
This would be much easier to read

Reply:
hmm that could work too. But honestly, the 2 methods* gets executed regardless, so not much difference.

* ``loadAllOnlinePlugins()`` is always executed first. ``afterPluginsLoadedEvent.invoke(true)`` runs on plugin update and download.